vm/actor/src/builtin/market/state.rs

```diff
@@ -0,0 +1,573 @@
+        self.locked_table = lt
+            .root()
+            .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?;
+        self.escrow_table = et
+            .root()
+            .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?;
+        Ok(())
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Deal state operations
+    ////////////////////////////////////////////////////////////////////////////////
+
+    #[allow(dead_code)]
+    pub(super) fn update_pending_deal_states_for_party<BS>(
+        &mut self,
+        store: &BS,
+        epoch: ChainEpoch,
+        addr: &Address,
+    ) -> Result<TokenAmount, ActorError>
+    where
+        BS: BlockStore,
+    {
+        // TODO check if rt curr_epoch can be 0
+        let epoch = epoch - 1;
+        let dbp = SetMultimap::from_root(store, &self.deal_ids_by_party)
+            .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e.into()))?;
+
+        let mut extracted_ids = Vec::new();
+        dbp.for_each(addr, |id| {
+            extracted_ids.push(id);
+            Ok(())
+        })
+        .map_err(|e| ActorError::new(ExitCode::ErrIllegalState, e))?;
+
+        self.update_pending_deal_states(store, extracted_ids, epoch)
+    }
+    #[allow(dead_code)]
+    pub(super) fn update_pending_deal_states<BS>(
+        &mut self,
+        store: &BS,
+        deal_ids: Vec<DealID>,
+        epoch: ChainEpoch,
+    ) -> Result<TokenAmount, ActorError>
+    where
+        BS: BlockStore,
+    {
+        let mut amount_slashed_total = TokenAmount::zero();
+
+        for deal in deal_ids {
+            amount_slashed_total += self.update_pending_deal_state(store, deal, epoch)?;
+        }
+
+        Ok(amount_slashed_total)
+    }
+    #[allow(dead_code)]
+    pub(super) fn update_pending_deal_state<BS>(
+        &mut self,
+        store: &BS,
+        deal_id: DealID,
+        epoch: ChainEpoch,
+    ) -> Result<TokenAmount, ActorError>
+    where
+        BS: BlockStore,
+    {
+        let deal = self.must_get_deal(store, deal_id)?;
+        let mut state = self.must_get_deal_state(store, deal_id)?;
+
+        let ever_updated = state.last_updated_epoch.is_some();
+        let ever_slashed = state.slash_epoch.is_some();
+
+        if ever_updated && state.last_updated_epoch.unwrap() > epoch {
+            return Err(ActorError::new(
+                ExitCode::ErrIllegalState,
+                "Deal was updated in the future".to_owned(),
+            ));
+        }
+
+        if state.sector_start_epoch.is_none() {
+            if epoch > deal.start_epoch {
+                return self.process_deal_init_timed_out(store, deal_id, deal, state);
+            }
+            return Ok(TokenAmount::zero());
+        }
+
+        assert!(deal.start_epoch <= epoch);
+
+        let deal_end = if ever_slashed {
+            assert!(state.slash_epoch.unwrap() <= deal.end_epoch);
```

Comment:
Do we want to leave assets in the code?

Reply:
asserts*

Reply:
no reason not to I don't think, if it fails it is a bug or something malicious, and would be caught. Not really a good reason to just pass back an error that could be missed

Reply:
True. Would it make sense to fail with a message then? Or do asserts give you the line number?

Reply:
They do, I can add a message anyway I guess, this isn't something we expect to ever hit though, just a sanity check